### PR TITLE
Bug 1787907: use the machine config uid for the kubeletconfig finalizer name

### PR DESCRIPTION
**Pull request by rphillips**
**- What I did**
This PR changes the KubeletConfig finalizer name to be the Machine Config UID. Finalizer names are restricted to 63 characters, so machine names impact if the kubelet config finalizer will be set.

The change preserves backwards compatibility by attempting the deletion of the MachineConfig by name, then falling back to Listing the machineconfigs and finding the UID that matches.

**- How to verify it**

**- Description for the changelog**
```
KubeletConfig finalizers have changed to using the MachineConfig UID.
```

### CODE REVIEW

**Comment by rphillips** [comment]
spelling issue fixed

File: pkg/controller/kubelet-config/kubelet_config_controller.go [hunk 1]
```diff
@@ -221,9 +221,25 @@ func (ctrl *Controller) cascadeDelete(cfg *mcfgv1.KubeletConfig) error {
 	if len(cfg.GetFinalizers()) == 0 {
 		return nil
 	}
-	mcName := cfg.GetFinalizers()[0]
-	err := ctrl.client.MachineconfigurationV1().MachineConfigs().Delete(mcName, &metav1.DeleteOptions{})
+	finalizerName := cfg.GetFinalizers()[0]
+	// for backwards compatibility try deleting the full Machine Config by name
```
Comment (ashcrow) [inline pkg/controller/kubelet-config/kubelet_config_controller.go]: nit: Won't this continue to be the case once UUIDs start being used _as_ the names themselves?

Comment (rphillips) [reply pkg/controller/kubelet-config/kubelet_config_controller.go]: ... yes... I force pushed a change to simplify the logic to list the machine configs and then check for ownership for the delete.

**Comment by ashcrow** [comment]
/test ci/prow/e2e-gcp-op

Looks like the MCD fell over in a way none of the other tests found. I'm guessing flake.

File: pkg/controller/kubelet-config/kubelet_config_controller.go [hunk 2]
```diff
@@ -221,11 +221,20 @@ func (ctrl *Controller) cascadeDelete(cfg *mcfgv1.KubeletConfig) error {
...
-	mcName := cfg.GetFinalizers()[0]
-	err := ctrl.client.MachineconfigurationV1().MachineConfigs().Delete(mcName, &metav1.DeleteOptions{})
-	if err != nil && !macherrors.IsNotFound(err) {
+	finalizerName := cfg.GetFinalizers()[0]
+	mcs, err := ctrl.client.MachineconfigurationV1().MachineConfigs().List(metav1.ListOptions{})
+	if err != nil {
 		return err
 	}
+	for _, mc := range mcs.Items {
+		if string(mc.GetUID()) == finalizerName || mc.GetName() == finalizerName {
```
Comment (ashcrow) [inline pkg/controller/kubelet-config/kubelet_config_controller.go]: Clarification question: I read this as:

    If the UID or the Name is the finalizerName then delete the MC by it's Name.

Is there any _reasonable_ possibility that the name _is_ the UID but us *not* the Name?

Comment (ashcrow) [reply pkg/controller/kubelet-config/kubelet_config_controller.go]: My guess is the answer is no, but I'd like to verify :slightly_smiling_face:

Comment (rphillips) [reply pkg/controller/kubelet-config/kubelet_config_controller.go]: No.... The delete API always uses the name. :smiley:

**Comment by rphillips** [comment]
weird errors I haven't seen before:
```
    mcd_test.go:388: machine config hasn't been picked by any MCD: timed out waiting for the condition
    mcd_test.go:402: worker pool isn't reporting degraded with a bad MC: timed out waiting for the condition
    mcd_test.go:412: pool worker didn't report updated to rendered-worker-0bc600603db20219d5e3323154561b93: timed out waiting for the condition
```

going to 
/retest

**Comment by rphillips** [comment]
@kikisdeliveryservice @sinnykumari friendly ping for a review.

**Comment by ashcrow** [comment]
We may not want to continue trying to pass e2e-gcp-op right now as it seems to not be passing in general.

```
level=error msg="Error: Error creating service account: googleapi: Error 429: Maximum number of service accounts on project reached., rateLimitExceeded"
```

**Comment by ashcrow** [comment]
e2e-gcp-op history shows it's rarely passing anywhere for what seems like infra reasons (service accounts, cpu limits, api, etc..)

**Comment by cgwalters** [comment]
> e2e-gcp-op history shows it's rarely passing anywhere for what seems like infra reasons (service accounts, cpu limits, api, etc..)

Right, but e2e-gcp-op is the MCO's test suite so...we need it to work.  I think we need someone diving into freeing up GCP resource.s

**Comment by ashcrow** [comment]
> > e2e-gcp-op history shows it's rarely passing anywhere for what seems like infra reasons (service accounts, cpu limits, api, etc..)
> 
> Right, but e2e-gcp-op is the MCO's test suite so...we need it to work. I think we need someone diving into freeing up GCP resource.s

cc @runcom

**Comment by ashcrow** [comment]
Putting on hold until someone can look at the MCO testing.

**Comment by ashcrow** [comment]
Looks like testing folks are on it.

**Pull request merged**